annotated.go

```diff
@@ -89,3 +90,268 @@ func (a Annotated) String() string {
+//
+//     Field1 T1 `$tags[0]`
+//     Field2 T2 `$tags[1]`
+//     ...
+//     FieldN TN `$tags[N-1]`
+//   }
+func (rt resultTags) getAnnotatedType(fType reflect.Type) []reflect.Type {
+	annotatedResult := []reflect.StructField{{
+		Name:      "Out",
+		Type:      reflect.TypeOf(Out{}),
+		Anonymous: true,
+	}}
+
+	errInterface := reflect.TypeOf((*error)(nil)).Elem()
+	numAnnotated := 0
+
+	for i := 0; i < fType.NumOut(); i++ {
+		// guard against error results
+		if fType.Out(i).Implements(errInterface) {
+			continue
+		}
+		structField := reflect.StructField{
+			Name: fmt.Sprintf("Field%d", numAnnotated),
+			Type: fType.Out(i),
+		}
+		if numAnnotated < len(rt.tags) {
+			structField.Tag = reflect.StructTag(rt.tags[numAnnotated])
+		}
+		numAnnotated++
+		annotatedResult = append(annotatedResult, structField)
+	}
+
+	resultTypes := []reflect.Type{reflect.StructOf(annotatedResult)}
+	return resultTypes
+}
+
+// ResultTags is an Annotation that annotates the result(s) of a function.
+// When multiple tags are specified, each tag is mapped to the corresponding
+// positional result.
+func ResultTags(tags ...string) Annotation {
+	r := resultTags{tags}
+	return r
+}
+
+// Annotate lets you annotate a function's paramter and returns with tags
+// without you having to declare separate struct definitions for them.
+//
+// For example,
+//   func NewGateway(ro, rw *db.Conn) *Gateway { ... }
+//   fx.Provide(
+//     fx.Annotate(
+//       NewGateway,
+//       fx.ParamTags(`name:"ro" optional:"true"`, `name:"rw"`),
+//       fx.ResultTags(`name:"foo"`),
+//     ),
+//   )
+//
+// Is equivalent to,
+//
+//  type params struct {
+//    fx.In
+//
+//    RO *db.Conn `name:"ro" optional:"true"`
+//    RW *db.Conn `name:"rw"`
+//  }
+//
+//  type result struct {
+//    fx.Out
+//
+//    GW *Gateway `name:"foo"`
+//   }
+//
+//   fx.Provide(func(p params) result {
+//     return result{GW: NewGateway(p.RO, p.RW)}
+//   })
+//
+// If a single annotation appears multiple times, all annotations
+// will be ignored.
+//
+// For example,
+//
+//  fx.Provide(
+//    fx.Annotate(
+//      NewGateWay,
+//      fx.ParamTags(`name:"ro" optional:"true"`),
+//      fx.ParamTags(`name:"rw"),
+//      fx.ResultTags(`name:"foo"`)
+//    )
+//  )
+//
+// is considered an invalid usage and will not apply any of the
+// Annotations to NewGateway.
+//
+// If more tags are given than the number of parameters/results, only
+// the ones up to the number of parameters/results will be applied.
+func Annotate(f interface{}, anns ...Annotation) interface{} {
+	fVal := reflect.ValueOf(f)
+	fType := fVal.Type()
+	numIn := fType.NumIn()
+	numOut := fType.NumOut()
+
+	var annotations annotations
+	var ins []reflect.Type
+	var outs []reflect.Type
+
+	annotatedIn := false
+	annotatedOut := false
+
+	for _, ann := range anns {
+		if e := ann.apply(&annotations); e != nil {
+			return annotationError{
+				target: f,
+				err:    e,
+			}
+		}
+		if pTags, ok := ann.(paramTags); ok {
```

Comment:
style nit: type switch might be more readable here.

```
switch ann := ann.(type) {
case paramTags:
  ann.getAnnotatedType(..)

case resultTags:
  ann.getAnnotatedType(..)

default:
  panic(...)
}
```

(I added a panic in this example because if we see an annotation that is not one of the two we support, there's a catastrophic bugs somewhere. There's a [blurb that Dig uses for panics](https://github.com/uber-go/dig/blob/b79edae8e13848e8ca4c84f54f8be680d5ba7605/result.go#L167-L172) like that that inform the user that this is a bug and that they should file a bug report. We can use something like that here.)